druntime/src/object.d

```diff
@@ -2614,6 +2622,13 @@ class Throwable : Object
     {
         if (nextInChain && nextInChain._refcount)
             _d_delThrowable(nextInChain);
+        // handle owned traceinfo
+        if (infoDeallocator !is null)
+        {
+            infoDeallocator(info);
+            info = null; // avoid any kind of dangling pointers if we can help
+                         // it.
+        }
```

Comment:
Out of curiousity, is there any reason why the deallocator pointer needs to stick around after `info` has been freed?  Could the deallocator be moved into the `TraceInfo` interface instead?

Reply:
Yes, because you aren't guaranteed that `info` still points at a valid object (GC can collect in any order).

On an earlier revision (before committing), I had done something like this.

Reply:
Just to clarify, i mean:
```
if (info.deallocator !is null)
{
    info.deallocator(info);
    info = null
}
```

If `info` no longer points at a valid object, surely then `infoDeallocator(info)` would crash regardless?

Reply:
The fact that `infoDeallocator` is not null indicates that the GC is not the owner (hence, deallocation is needed). For cases where `info` *is* GC allocated, running a virtual function on it might crash randomly depending on the collection order. Surely, when the exception is garbage, the traceinfo also is.

One other possibility is to store a boolean indicating that the info is not GC allocated, and then the traceinfo should be deallocated via a call to some virtual function inside it, and in fact, I did do this originally. But I was concerned about a) having to implement the deallocation into every implementing class of TraceInfo, and b) how deallocating the class during a member function might interact with a class invariant.

So I picked this way instead. And actually, I think it's better, because really the issue was the allocation of the class itself, not the innards.

And.... thinking about this a bit more, I should change InvalidMemoryOperationError and others to directly call the default trace info function, instead of hoping the user's custom function does it properly.